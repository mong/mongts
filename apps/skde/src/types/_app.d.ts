export {};

declare global {
  interface Window {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    _mtm: any;
  }
}
