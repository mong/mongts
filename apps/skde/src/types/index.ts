export type { AtlasData } from "./AtlasData";
export type { AnalyseData, Tag } from "./AnalyseTypes";
