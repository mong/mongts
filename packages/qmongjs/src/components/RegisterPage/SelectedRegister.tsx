import { useState, useRef, useEffect } from "react";
import { useQueryParam } from "use-query-params";
import { UseQueryResult } from "@tanstack/react-query";

import SelectTreatmentUnits from "../SelectTreatmentUnits";
import SelectYear from "../SelectYear";

import { Header } from "./header";
import styles from "./registerPage.module.css";

import {
  mainQueryParamsConfig,
  maxYear,
  minYear,
  defaultYear,
} from "../../app_config";

import {
  useResizeObserver,
  useUnitNamesQuery,
  useSelectionYearsQuery,
  useIndicatorQuery,
} from "../../helpers/hooks";
import { mathClamp, validateTreatmentUnits } from "../../helpers/functions";
import { UnitNameList } from "./unitnamelist";
import { NestedTreatmentUnitName } from "./unitnamelist/unitnamelistbody";
import { RegisterName, OptsTu, Indicator } from "types";

import LEGEND from "../TargetLevels";
import { IndicatorTable } from "../IndicatorTable";
import { useRouter } from "next/router";

interface SelectedRegisterProps {
  registerNames: RegisterName[];
}

export const SelectedRegister = ({ registerNames }: SelectedRegisterProps) => {
  const { register, tab } = useRouter().query as {
    register: string;
    tab: string;
  };
  const tabNames: { label: string; value: string }[] = [
    { value: "sykehus", label: "Sykehus" },
    { value: "opptaksomraade", label: "Opptaksområde" },
    { value: "datakvalitet", label: "Datakvalitet" },
  ];
  const context =
    tab === "sykehus"
      ? "caregiver"
      : tab === "opptaksomraade"
        ? "resident"
        : tab === "datakvalitet"
          ? "coverage"
          : "caregiver";
  const registerInfo = registerNames.filter((reg) => reg.rname === register);

  const queryContext =
    context === "coverage"
      ? { context: "caregiver", type: "dg" }
      : { context, type: "ind" };

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const unitNamesQuery: UseQueryResult<any, unknown> = useUnitNamesQuery(
    register as string,
    queryContext.context,
    queryContext.type,
  );
  const nestedUnitNames: NestedTreatmentUnitName[] | [] =
    unitNamesQuery.data?.nestedUnitNames ?? [];
  const optstu: OptsTu[] | [] = unitNamesQuery.data?.opts_tu ?? [];

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const selectionYearQuery: UseQueryResult<any, unknown> =
    useSelectionYearsQuery(
      register as string,
      queryContext.context,
      queryContext.type,
    );

  const valid_years =
    selectionYearQuery.data ??
    Array.from(Array(maxYear - minYear + 1).keys()).map((v) => minYear + v);

  const [selection_bar_height, update_selection_bar_height] = useState<
    number | null
  >(null);
  const [legend_height, update_legend_height] = useState(null);
  //height of the selection bar
  const selection_bar_ref = useRef<HTMLDivElement | null>(null);
  const selection_bar_dim = useResizeObserver(selection_bar_ref);
  useEffect(() => {
    if (!selection_bar_dim) {
      return;
    }
    const top = (selection_bar_dim.target as HTMLElement).offsetHeight ?? null;
    update_selection_bar_height(top);
  }, [selection_bar_dim]);

  const [treatment_units, update_treatment_units] = useQueryParam(
    "selected_treatment_units",
    mainQueryParamsConfig.selected_treatment_units,
  );

  const [selected_year, update_selected_year] = useQueryParam(
    "year",
    mainQueryParamsConfig.year,
  );
  const min = valid_years.length === 0 ? minYear : Math.min(...valid_years);
  const max = valid_years.length === 0 ? maxYear : Math.max(...valid_years);

  const lastCompleteYear = (data: Indicator[] | undefined) => {
    if (!data) {
      return defaultYear;
    }
    // get the last year with complete data
    const latestAffirmDates = data.flatMap((d) =>
      d.delivery_latest_affirm
        ? new Date(d.delivery_latest_affirm).getTime()
        : new Date(defaultYear + 1, 0).getTime(),
    );
    const lastCompleteYear =
      new Date(
        Math.max(...latestAffirmDates) +
          // add a couple of days so 30. des 1980 will return 1980 instead of 1979.
          48 * 60 * 60 * 1000,
      ).getFullYear() - 1;
    return lastCompleteYear;
  };

  const { data } = useIndicatorQuery({
    registerShortName: register,
    context: queryContext.context,
    type: queryContext.type,
  });

  const validated_selected_year = mathClamp(
    selected_year || lastCompleteYear(data),
    min,
    max,
  );

  const validated_treatment_units = validateTreatmentUnits(
    treatment_units as string[],
    optstu,
  );

  const colspan = validated_treatment_units.length + 2;

  const [show_level_filter, update_show_level_filter] = useQueryParam<
    string | undefined
  >("level", mainQueryParamsConfig.level);

  const validReg = registerNames.map((d) => d.rname);
  // returns page not found if the register param or tab param is not valid
  if (
    !validReg.includes(register as string) ||
    !tabNames.some((tabName) => tabName.value === tab) ||
    (!registerInfo[0].caregiver_data && tab === "sykehus") ||
    (!registerInfo[0].resident_data && tab === "opptaksomraade") ||
    (!registerInfo[0].dg_data && tab === "datakvalitet")
  ) {
    return (
      <div style={{ minHeight: "100vh" }}>
        <h1 style={{ margin: "10%" }}>Page Not Found</h1>
      </div>
    );
  }
  const registerFullName = registerNames.filter((d) => d.rname === register)[0]
    .full_name;

  const registerTabs = tabNames.filter((tabs) => {
    if (registerInfo[0].caregiver_data && tabs.label === "Sykehus") {
      return true;
    }
    if (registerInfo[0].resident_data && tabs.label === "Opptaksområde") {
      return true;
    }
    if (registerInfo[0].dg_data && tabs.label === "Datakvalitet") {
      return true;
    }
    return false;
  });

  const placeholder =
    tab === "opptaksomraade" ? (
      <div>
        <i className="fas fa-search" /> Søk etter opptaksområder
      </div>
    ) : (
      <div>
        <i className="fas fa-search" /> Søk etter behandlingsenheter
      </div>
    );

  return (
    <div className={styles.appContainer} style={{ minHeight: "100vh" }}>
      <Header
        registerNames={registerNames}
        dataFrom={registerFullName}
        tabNames={registerTabs}
        activeTab={tab as string}
      />
      <div className={styles.appBody}>
        <div className={styles.selectionContainer} ref={selection_bar_ref}>
          <div className={styles.treatmentUnitSelection}>
            <SelectTreatmentUnits
              opts={optstu}
              update_tu={update_treatment_units}
              treatment_unit={validated_treatment_units}
              placeholder={placeholder}
            />
            <UnitNameList
              nestedUnitNames={nestedUnitNames}
              treatment_units={validated_treatment_units}
              update_treatment_units={update_treatment_units}
              multiple_choice={true}
            />
          </div>
          <div className={styles.yearSelection}>
            <SelectYear
              opts={valid_years.sort()}
              update_year={update_selected_year}
              selected_year={validated_selected_year}
            />
          </div>
        </div>

        <LEGEND
          update_show_level_filter={update_show_level_filter}
          show_level_filter={show_level_filter}
          selection_bar_height={selection_bar_height}
          update_legend_height={update_legend_height}
          width="100%"
        />

        <div className={styles.contentContainer}>
          <div className={styles.mainTableContainer}>
            <IndicatorTable
              context={context}
              tableType={"singleRegister"}
              registerNames={registerInfo}
              unitNames={[...validated_treatment_units, "Nasjonalt"]}
              treatmentYear={validated_selected_year}
              colspan={colspan}
              medicalFieldFilter={[register as string]}
              showLevelFilter={show_level_filter ?? ""}
              selection_bar_height={selection_bar_height ?? 0}
              legend_height={legend_height ?? 0}
            />
          </div>
        </div>
      </div>
    </div>
  );
};

export default SelectedRegister;
