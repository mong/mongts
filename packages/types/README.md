# types

Types shared among the packages (`api`, `skde` and `qmongjs`).
