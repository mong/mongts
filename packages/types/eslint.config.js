export default [
  {
    files: ["**/*.ts", "**/*.tsx"],
  },
];
