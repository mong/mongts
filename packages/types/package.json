{
  "name": "types",
  "version": "0.0.0",
  "main": "./src/index",
  "types": "./src/index",
  "packageManager": "yarn@4.2.1",
  "private": true,
  "devDependencies": {
    "@typescript-eslint/eslint-plugin": "7.8.0",
    "@typescript-eslint/parser": "7.8.0"
  }
}
