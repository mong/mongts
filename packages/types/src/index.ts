export interface Description {
  id: string;
  dg_id: string | null;
  include: number | null;
  title: string | null;
  name: string | null;
  type: string | null;
  sformat: string;
  measure_unit: string | null;
  min_denominator: number | null;
  min_value: number | null;
  max_value: number | null;
  level_green: number | null;
  level_yellow: number | null;
  level_direction: number | null;
  short_description: string | null;
  long_description: string | null;
  registry_id: number;
  rname: string | null;
  full_name: string;
}

export interface Indicator {
  id?: number;
  ind_id: string;
  unit_level: string;
  unit_name: string;
  context: "caregiver" | "resident";
  year: number;
  denominator: number;
  var: number;
  level_direction: number | null;
  level_green: number | null;
  level_yellow: number | null;
  sformat: string | null;
  dg: number | null;
  delivery_time?: Date | null;
  delivery_latest_update?: Date | null;
  delivery_latest_affirm?: Date | null;
  type?: string | null;
  include: number | null;
  min_denominator: number | null;
  registry_id: number;
  registry_name: string;
  registry_full_name: string;
  registry_short_name: string;
  medfield_id: number;
  medfield_name: string;
  medfield_full_name: string;
  ind_title: string | null;
}

export interface TuName {
  hospital: string;
  hf: string;
  hf_full: string;
  hf_shortest: string;
  hf_sort: number;
  rhf: string;
  rhf_shortest: string;
  rhf_sort: number;
}

export interface NestedTreatmentUnitName {
  rhf: string;
  rhf_sort: number;
  rhf_shortest: string;
  hf: {
    hf: string;
    hf_full: string;
    hf_shortest: string;
    hf_sort: number;
    hospital: string[];
  }[];
}

export interface OptsTu {
  label: "Sykehus" | "HF" | "RHF";
  options: {
    value: string;
    label: string;
  }[];
}

export interface RegisterName {
  id: number;
  rname: string;
  full_name: string;
  short_name: string;
  caregiver_data: 0 | 1 | null;
  resident_data: 0 | 1 | null;
  dg_data: 0 | 1 | null;
  url: string | null;
  description: string | null;
}

export type DataPoint = {
  id: number;
  unitName: string;
  year: number;
  var: number | null;
  denominator: number;
  dg: number | null;
  context: "caregiver" | "resident";
  deliveryTime: Date;
  affirmTime: Date;
  indicatorID: string;
};

export type IndicatorInfo = {
  indicatorID: string;
  indicatorTitle: string | null;
  levelGreen: number | null;
  levelYellow: number | null;
  levelDirection: number | null;
  minDenominator: number | null;
  minValue: number | null;
  maxValue: number | null;
  shortDescription: string | null;
  longDescription: string | null;
  indType:
    | "andel"
    | "beregnet_andel"
    | "antall"
    | "gjennomsnitt"
    | "median"
    | "dg_andel";
  sortingName: string | null;
  format: string | null;
  registerID: number;
};

export type IndicatorData = IndicatorInfo & { data?: DataPoint[] };

export type Registry = {
  registerFullName: string;
  registerName: string;
  registerShortName: string;
  registerID: number;
  medfieldID: number[];
};

export type Medfield = {
  name: string;
  registers: string[];
  shortName: string;
};

export type RegisterData = Registry & {
  indicatorData: IndicatorData[];
};

export type URLs = {
  shortName: string;
  url: string;
};

export type RegistryRank = {
 id: number;
 name: string;
 full_name: string;
 short_name: string;
 verdict: string;
 year: number;
}