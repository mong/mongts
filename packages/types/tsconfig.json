{
    "compilerOptions": {
      "target": "es5",
      "lib": ["dom", "dom.iterable", "esnext"],
      "allowJs": true,
      "strict": true
    },
    "include": ["src"]
}  