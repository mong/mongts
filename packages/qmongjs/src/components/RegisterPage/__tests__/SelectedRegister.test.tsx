import { QueryParamProvider } from "use-query-params";
import { QueryClient, QueryClientProvider } from "@tanstack/react-query";
import { render, screen, waitFor } from "@testing-library/react";

import opts from "../../../dev-tools/data/unitnames.json";
import { SelectedRegister } from "../SelectedRegister";
import { RegisterName } from "types";

import mockRouter from "next-router-mock";
import { NextAdapter } from "next-query-params";
import { createDynamicRouteParser } from "next-router-mock/dynamic-routes";
import {
  useUnitNamesQuery,
  useSelectionYearsQuery,
  useIndicatorQuery,
} from "../../../helpers/hooks";
import { buildIndicator } from "../../../test/test_data/dataBuilder";
import { useRouter } from "next/router";

jest.mock("next/router", () => require("next-router-mock"));
// This is needed for mocking 'next/link':
jest.mock("next/dist/client/router", () => require("next-router-mock"));
jest.mock("../../../helpers/hooks");
jest.mock("next/router", () => ({
  useRouter: jest.fn(),
}));

mockRouter.useParser(
  createDynamicRouteParser([
    // These paths should match those found in the `/pages` folder:
    "/hoftebrudd/[tab].js",
    "/[register]/[tab].js",
  ])
);
beforeEach(() => {
  mockRouter.setCurrentUrl("/hoftebrudd/sykehus");
});

const registerInfo: RegisterName[] = [
  {
    id: 1,
    rname: "hoftebrudd",
    full_name: "Nasjonalt hoftebruddregister",
    caregiver_data: 1,
    resident_data: 0,
    dg_data: 1,
    url: null,
    description: null,
  },
];
it("SelectedRegister renders correctly for a single registry", async () => {
  (useUnitNamesQuery as jest.Mock).mockReturnValue({
    data: {
      nestedUnitNames: [{ rhf: "Helse 1", hf: ["Helse 2"] }],
      opts_tu: [
        { label: "rhf", options: ["Helse 1"] },
        { label: "hf", options: ["Helse 2"] },
      ],
    },
    isLoading: false,
    error: false,
  });
  (useSelectionYearsQuery as jest.Mock).mockReturnValue({
    data: [2019, 2020, 2021, 2022],
    isLoading: false,
    error: false,
  });

  (useIndicatorQuery as jest.Mock).mockReturnValue({
    data: [buildIndicator({})],
    isLoading: false,
    error: false,
  });

  (useRouter as jest.Mock).mockReturnValue({
    query: { register: "hoftebrudd", tab: "sykehus" },
  });

  const queryClient = new QueryClient();

  const { container } = render(
    //    <QueryClientProvider client={queryClient}>
    <QueryParamProvider adapter={NextAdapter}>
      <SelectedRegister registerNames={registerInfo} />
    </QueryParamProvider>
    //    </QueryClientProvider>
  );

  //  expect(container).toMatchSnapshot();
});
