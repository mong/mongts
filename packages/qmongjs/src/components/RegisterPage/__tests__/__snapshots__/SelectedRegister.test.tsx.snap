// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectedRegister renders correctly for a single registry 1`] = `
<div>
  <div
    style="min-height: 100vh;"
  >
    <h1
      style="margin: 10%;"
    >
      Page Not Found
    </h1>
  </div>
</div>
`;
